cannon/mipsevm/exec/mips_instructions.go

```diff
@@ -1,96 +1,114 @@
 package exec
 
 import (
+	"fmt"
+
 	"github.com/ethereum-optimism/optimism/cannon/mipsevm"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/arch"
 	"github.com/ethereum-optimism/optimism/cannon/mipsevm/memory"
+
+	// TODO: MIPS64 port. Replace with a custom library
+	u128 "lukechampine.com/uint128"
 )
 
 const (
 	OpLoadLinked       = 0x30
 	OpStoreConditional = 0x38
 )
 
-func GetInstructionDetails(pc uint32, memory *memory.Memory) (insn, opcode, fun uint32) {
+func GetInstructionDetails(pc Word, memory *memory.Memory) (insn, opcode, fun uint32) {
 	insn = memory.GetMemory(pc)
 	opcode = insn >> 26 // First 6-bits
 	fun = insn & 0x3f   // Last 6-bits
 
 	return insn, opcode, fun
 }
 
-func ExecMipsCoreStepLogic(cpu *mipsevm.CpuScalars, registers *[32]uint32, memory *memory.Memory, insn, opcode, fun uint32, memTracker MemTracker, stackTracker StackTracker) (memUpdated bool, memAddr uint32, err error) {
+func ExecMipsCoreStepLogic(cpu *mipsevm.CpuScalars, registers *[32]Word, memory *memory.Memory, insn, opcode, fun uint32, memTracker MemTracker, stackTracker StackTracker) (memUpdated bool, memAddr Word, err error) {
 	// j-type j/jal
 	if opcode == 2 || opcode == 3 {
-		linkReg := uint32(0)
+		linkReg := Word(0)
 		if opcode == 3 {
 			linkReg = 31
 		}
 		// Take top 4 bits of the next PC (its 256 MB region), and concatenate with the 26-bit offset
-		target := (cpu.NextPC & 0xF0000000) | ((insn & 0x03FFFFFF) << 2)
+		target := (cpu.NextPC & SignExtend(0xF0000000, 32)) | Word((insn&0x03FFFFFF)<<2)
 		stackTracker.PushStack(cpu.PC, target)
 		err = HandleJump(cpu, registers, linkReg, target)
 		return
 	}
 
 	// register fetch
-	rs := uint32(0) // source register 1 value
-	rt := uint32(0) // source register 2 / temp value
-	rtReg := (insn >> 16) & 0x1F
+	rs := Word(0) // source register 1 value
+	rt := Word(0) // source register 2 / temp value
+	rtReg := Word((insn >> 16) & 0x1F)
 
 	// R-type or I-type (stores rt)
 	rs = registers[(insn>>21)&0x1F]
 	rdReg := rtReg
 	if opcode == 0 || opcode == 0x1c {
 		// R-type (stores rd)
 		rt = registers[rtReg]
-		rdReg = (insn >> 11) & 0x1F
+		rdReg = Word((insn >> 11) & 0x1F)
 	} else if opcode < 0x20 {
 		// rt is SignExtImm
 		// don't sign extend for andi, ori, xori
 		if opcode == 0xC || opcode == 0xD || opcode == 0xe {
 			// ZeroExtImm
-			rt = insn & 0xFFFF
+			rt = Word(insn & 0xFFFF)
 		} else {
 			// SignExtImm
-			rt = SignExtend(insn&0xFFFF, 16)
+			rt = SignExtendImmediate(insn)
 		}
 	} else if opcode >= 0x28 || opcode == 0x22 || opcode == 0x26 {
 		// store rt value with store
 		rt = registers[rtReg]
 
 		// store actual rt with lwl and lwr
 		rdReg = rtReg
+	} else if opcode == 0x27 || opcode == 0x1A || opcode == 0x1B { // 64-bit opcodes
```

Comment:
Do we need to make sure these are disabled for 32-bit mode? It's possible that's handled elsewhere.

Reply:
It's handled elsewhere by `ExecuteMipsInstruction`, but I've added an assert here anyways to be defensive.